File: Stack_Linked_List.c
// Menu driven program for linked list implementation of stack :-

#include <stdio.h>
#include <stdlib.h>

// Node structure
struct Node {
    int data;
    struct Node* next;
};

// Function to push an element onto the stack
void push(struct Node** top, int value) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    if (!newNode) {
        printf("Memory allocation failed!\n");
        return;
    }
    newNode->data = value;
    newNode->next = *top;
    *top = newNode;
    printf("%d pushed to stack\n", value);
}

// Function to pop an element from the stack
int pop(struct Node** top) {
    if (*top == NULL) {
        printf("Stack underflow!\n");
        return -1;
    }
    struct Node* temp = *top;
    int poppedValue = temp->data;
    *top = (*top)->next;
    free(temp);
    return poppedValue;
}

// Function to display the stack
void display(struct Node* top) {
    if (top == NULL) {
        printf("Stack is empty!\n");
        return;
    }
    struct Node* temp = top;
    printf("Stack elements:\n");
    while (temp) {
        printf("%d -> ", temp->data);
        temp = temp->next;
    }
    printf("NULL\n");
}

int main() {
    struct Node* top = NULL;
    int choice, value;

    while (1) {
        printf("\nStack Menu (Linked List Implementation):\n");
        printf("1. Push\n2. Pop\n3. Display\n4. Exit\n");
        printf("Enter your choice: ");
        scanf("%d", &choice);

        switch (choice) {
            case 1:
                printf("Enter value to push: ");
                scanf("%d", &value);
                push(&top, value);
                break;
            case 2:
                value = pop(&top);
                if (value != -1)
                    printf("Popped value: %d\n", value);
                break;
            case 3:
                display(top);
                break;
            case 4:
                printf("Exiting...\n");
                exit(0);
            default:
                printf("Invalid choice! Try again.\n");
        }
    }
    return 0;
}

File: Stack_Linked_List.py
# Menu driven program for linked list implementation of stack :-

class Node:
    """Class representing a Node in the linked list"""
    def __init__(self, data):
        self.data = data
        self.next = None

class Stack:
    """Class representing Stack using a linked list"""
    def __init__(self):
        self.top = None

    def push(self, value):
        """Pushes an element onto the stack"""
        new_node = Node(value)
        new_node.next = self.top
        self.top = new_node
        print(f"{value} pushed to stack")

    def pop(self):
        """Pops an element from the stack"""
        if self.top is None:
            print("Stack underflow!")
            return None
        popped_value = self.top.data
        self.top = self.top.next
        print(f"Popped value: {popped_value}")
        return popped_value

    def display(self):
        """Displays the stack"""
        if self.top is None:
            print("Stack is empty!")
            return
        temp = self.top
        print("Stack elements:")
        while temp:
            print(temp.data, end=" -> ")
            temp = temp.next
        print("NULL")

def menu():
    """Menu-driven function to use the stack"""
    stack = Stack()
    
    while True:
        print("\nStack Menu (Linked List Implementation):")
        print("1. Push\n2. Pop\n3. Display\n4. Exit")
        choice = int(input("Enter your choice: "))

        if choice == 1:
            value = int(input("Enter value to push: "))
            stack.push(value)
        elif choice == 2:
            stack.pop()
        elif choice == 3:
            stack.display()
        elif choice == 4:
            print("Exiting...")
            break
        else:
            print("Invalid choice! Try again.")

# Run the menu-driven program
if __name__ == "__main__":
    menu()
